### PR TITLE
chore: improve chart usage update query

### PR DESCRIPTION
**Pull request by almeidabbm**
Closes: 

### Description:

- Updates the chart usage taking into account field type
- Improves the chart usage update query so that it only performs at most 2 queries, one for increments and one for decrements

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/models/CatalogModel/CatalogModel.ts [hunk 1]
```diff
@@ -442,36 +444,78 @@ export class CatalogModel {
...
+                        });
+                    })
+                    .andWhere(
+                        `${CatalogTableName}.cached_explore_uuid`,
+                        cachedExploreUuid,
+                    )
+                    .andWhere(`${CatalogTableName}.project_uuid`, projectUuid)
+                    .andWhere('chart_usage', '>', 0) // Ensure we don't decrement below 0
+                    .decrement('chart_usage', 1),
             );
```
Comment (almeidabbm) [inline packages/backend/src/models/CatalogModel/CatalogModel.ts:519]:
```
update "catalog_search" set "chart_usage" = "chart_usage" - ? where (("catalog_search"."name" in (?) and "catalog_search"."field_type" = ?) or ("catalog_search"."name" in (?) and "catalog_search"."field_type" = ?)) and "catalog_search"."cached_explore_uuid" = ? and "catalog_search"."project_uuid" = ? and "chart_usage" > ?
```

File: packages/backend/src/models/CatalogModel/CatalogModel.ts [hunk 2]
```diff
@@ -442,36 +444,78 @@ export class CatalogModel {
...
+                                FieldType.DIMENSION,
+                            );
+                        });
+                    })
+                    .andWhere(
+                        `${CatalogTableName}.cached_explore_uuid`,
+                        cachedExploreUuid,
+                    )
+                    .andWhere(`${CatalogTableName}.project_uuid`, projectUuid)
+                    .increment('chart_usage', 1),
```
Comment (almeidabbm) [inline packages/backend/src/models/CatalogModel/CatalogModel.ts:484]:
```
update "catalog_search" set "chart_usage" = "chart_usage" + ? where ((1 = ? and "catalog_search"."field_type" = ?) or ("catalog_search"."name" in (?) and "catalog_search"."field_type" = ?)) and "catalog_search"."cached_explore_uuid" = ? and "catalog_search"."project_uuid" = ?
```

**Comment by joaoviana** [comment]
before
Screenshot 2024-12-10 at 09 21 49


after
Screenshot 2024-12-10 at 09 22 00


after adding a new chart that uses unique payment count

image


after refreshing catalog, it remains the same

image

File: packages/backend/src/models/CatalogModel/utils/index.ts [hunk 3]
```diff
@@ -117,85 +119,154 @@ export async function getCatalogFieldWhereByFieldIds(
...
+            metrics: removedMetrics,
+        },
+    };
+}
+
+export function getCatalogFieldWhereStatements(
+    fieldIds: string[],
+    fieldTableNameMap: Record<string, string | undefined>,
+    cachedExploreUuidTableNameMap: Record<string, string>,
+    fieldsType: FieldType,
```
Comment (joaoviana) [inline packages/backend/src/models/CatalogModel/utils/index.ts:161]:
```suggestion
    fieldType: FieldType,
```

NP ⬆️

**Review by joaoviana — APPROVED** [review]

File: packages/backend/src/models/CatalogModel/CatalogModel.ts [hunk 4]
```diff
@@ -432,49 +432,89 @@ export class CatalogModel {
...
                         )
-                        .where('chart_usage', '>', 0) // Ensure we don't decrement below 0
+                        .andWhere('chart_usage', '>', 0) // Ensure we don't decrement below 0
                         .decrement('chart_usage', 1),
-            );
+                );
+            }
 
-            await Promise.all([...incrementPromises, ...decrementPromises]);
+            await Promise.all(transactions);
```
Comment (almeidabbm) [inline packages/backend/src/models/CatalogModel/CatalogModel.ts:517]: @joaoviana Updated the code so it runs at most 2 transactions, 1 to increment and one to decrement

here's the resulting sql:

Increment
```sql
update "catalog_search" set "chart_usage" = "chart_usage" + ? where (("catalog_search"."cached_explore_uuid" = ? and "catalog_search"."name" = ? and "catalog_search"."field_type" = ?) or ("catalog_search"."cached_explore_uuid" = ? and "catalog_search"."name" = ? and "catalog_search"."field_type" = ?)) and "catalog_search"."project_uuid" = ?
```

Decrement
```sql
update "catalog_search" set "chart_usage" = "chart_usage" - ? where (("catalog_search"."cached_explore_uuid" = ? and "catalog_search"."name" = ? and "catalog_search"."field_type" = ?) or ("catalog_search"."cached_explore_uuid" = ? and "catalog_search"."name" = ? and "catalog_search"."field_type" = ?)) and "catalog_search"."project_uuid" = ? and "chart_usage" > ?
```

In both cases I added/removed 2 metrics so we could see the `OR` statements

**Review by joaoviana — APPROVED** [review]

**Pull request merged**